### PR TITLE
detect-detection: Convert unittests to new FAIL/PASS API

### PR DESCRIPTION
**Pull request by hsadia538**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Previous PR: #8083
Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4019) ticket:

Draft to gather feedback

Describe changes:

- Update detect-detection unit tests to FAIL/PASS APIs

### CODE REVIEW

**Comment by hsadia538** [comment]
I havent yet resolved all the review comments but I am getting the following when I run the test:
Screenshot 2022-10-25 at 11 37 59 PM

If I break the checks on `line 308` I get a segmentation fault and the remaining tests do not run.See below for reference
Screenshot 2022-10-25 at 11 42 22 PM

File: src/detect-detection-filter.c [hunk 1]
```diff
@@ -377,12 +378,13 @@ static int DetectDetectionFilterTestParse06 (void)
 {
     DetectThresholdData *df = NULL;
     df = DetectDetectionFilterParse("count 10, track by_dst, seconds 0");
-    if (df && (df->track == TRACK_DST) && (df->count == 10) && (df->seconds == 0)) {
-        DetectDetectionFilterFree(NULL, df);
-        return 0;
-    }
+    FAIL_IF_NOT_NULL(df);
```
Comment (victorjulien) [inline src/detect-detection-filter.c:381]: this fails if df != NULL, meaning that on the following line... df == NULL

**Comment by jufajardini** [comment]
> I havent yet resolved all the review comments but I am getting the following when I run the test: Screenshot 2022-10-25 at 11 37 59 PM
> 
> If I break the checks on `line 308` I get a segmentation fault and the remaining tests do not run.See below for reference Screenshot 2022-10-25 at 11 42 22 PM

From your screenshots, it seems that you're getting the segmentation fault at bit randomly?

> I havent yet resolved all the review comments but I am getting the following when I run the test: Screenshot 2022-10-25 at 11 37 59 PM
> 
> If I break the checks on `line 308` I get a segmentation fault and the remaining tests do not run.See below for reference Screenshot 2022-10-25 at 11 42 22 PM

Time to enable address sanitizer checks! \o\ (I'll do that, at least, as this made me curious...)

File: src/detect-detection-filter.c (hunk 1 above)
Comment (jufajardini) [reply src/detect-detection-filter.c:381]: 🙇🏽

Comment (hsadia538) [reply src/detect-detection-filter.c:381]: So just adding the line ` FAIL_IF_NOT_NULL(df);`  is enough and rest should be removed.?

**Comment by hsadia538** [comment]
> > I havent yet resolved all the review comments but I am getting the following when I run the test: Screenshot 2022-10-25 at 11 37 59 PM
> > If I break the checks on `line 308` I get a segmentation fault and the remaining tests do not run.See below for reference Screenshot 2022-10-25 at 11 42 22 PM
> 
> From your screenshots, it seems that you're getting the segmentation fault at bit randomly?
> 
> > I havent yet resolved all the review comments but I am getting the following when I run the test: Screenshot 2022-10-25 at 11 37 59 PM
> > If I break the checks on `line 308` I get a segmentation fault and the remaining tests do not run.See below for reference Screenshot 2022-10-25 at 11 42 22 PM
> 
> Time to enable address sanitizer checks! \o\ (I'll do that, at least, as this made me curious...)

In the first screenshot, all tests pass although we get an error(the same error is logged if I run the test on master)
In the second one, I think it happens in `DetectDetectionFilterTestParse02` that is where I broke the lengthy chunk into individual statements.

File: src/detect-detection-filter.c (hunk 1 above)
Comment (jufajardini) [reply src/detect-detection-filter.c:381]: I'd guess so.

Comment (hsadia538) [reply src/detect-detection-filter.c:381]: let me test again with this. Thanks

File: src/detect-detection-filter.c [hunk 2]
```diff
@@ -409,16 +410,12 @@ static int DetectDetectionFilterTestSig1(void)
...
-        goto end;
-    }
+    FAIL_IF_NULL(de_ctx);
 
     de_ctx->flags |= DE_QUIET;
 
     s = de_ctx->sig_list = SigInit(de_ctx,"alert tcp any any -> any 80 (msg:\"detection_filter Test\"; detection_filter: track by_dst, count 4, seconds 60; sid:1;)");
-    if (s == NULL) {
-        goto end;
-    }
```
Comment (jufajardini) [inline src/detect-detection-filter.c:421]: Here you can incorporate the feedback about replacing `SigInit` by `DetectEngineAppendSig`

File: src/detect-detection-filter.c [hunk 3]
```diff
@@ -440,19 +437,18 @@ static int DetectDetectionFilterTestSig1(void)
     SigMatchSignatures(&th_v, de_ctx, det_ctx, p);
     alerts += PacketAlertCheck(p, 1);
 
-    if(alerts == 4)
-        result = 1;
+    FAIL_IF_NOT(alerts == 4);
 
     SigGroupCleanup(de_ctx);
     SigCleanSignatures(de_ctx);
```
Comment (jufajardini) [inline src/detect-detection-filter.c:443]: When you use `DetectEngineAppendSig`, you can remove these.

File: src/detect-detection-filter.c (hunk 1 above)
Comment (hsadia538) [reply src/detect-detection-filter.c:381]: But in the scenario that fail is not raised due to df being not null we are not actually checking the rest.

Comment (jufajardini) [reply src/detect-detection-filter.c:381]: Hm. My understanding was: `DetectDetectionFilterParse` will return `NULL` if we have invalid values in `detection_filter`. And this test is described as a test for invalid values. Therefore... If `df` is not `NULL`, something went wrong and the test should fail. And as Victor said, if it is `NULL`, then we wouldn't have anything to check 🤔  

But maybe I understood it all wrongly...

File: src/detect-detection-filter.c (hunk 3 above)
Comment (victorjulien) [reply src/detect-detection-filter.c:443]: Technically unrelated, but please remove anyway :)

File: src/detect-detection-filter.c (hunk 1 above)
Comment (hsadia538) [reply src/detect-detection-filter.c:381]: Ok makes sense now,
If df is not null then it should fail regardless of something else and if it is null then we don't have anything to check. got it. 
Read the comment multiple times to make sense of this null vs not null confusion.

**Comment by hsadia538** [comment]
Followed by #8105

File: src/detect-detection-filter.c (hunk 1 above)
Comment (jufajardini) [reply src/detect-detection-filter.c:381]: I know the feeling xD

Comment (hsadia538) [reply src/detect-detection-filter.c:381]: trying to learn with every review but still a long way to go :'')

**Pull request closed without merging**